aws/resource_aws_traffic_mirror_filter_rule.go

```diff
@@ -0,0 +1,369 @@
+		} else {
+			//Modify request that adds port range seems to fail if protocol is not set in the request
+			input.Protocol = aws.Int64(int64(d.Get("protocol").(int)))
+			input.SetDestinationPortRange(buildTrafficMirrorPortRangeRequest(n))
+		}
+	}
+
+	if d.HasChange("source_port_range") {
+		_, v := d.GetChange("source_port_range")
+		n := v.([]interface{})
+		if 0 == len(n) {
+			removeFields = append(removeFields, aws.String("source-port-range"))
+		} else {
+			//Modify request that adds port range seems to fail if protocol is not set in the request
+			input.Protocol = aws.Int64(int64(d.Get("protocol").(int)))
+			input.SetSourcePortRange(buildTrafficMirrorPortRangeRequest(n))
+		}
+	}
+
+	if d.HasChange("rule_action") {
+		input.RuleAction = aws.String(d.Get("rule_action").(string))
+	}
+
+	if d.HasChange("rule_number") {
+		input.RuleNumber = aws.Int64(int64(d.Get("rule_action").(int)))
+	}
+
+	if d.HasChange("traffic_direction") {
+		input.TrafficDirection = aws.String(d.Get("traffic_direction").(string))
+	}
+
+	if len(removeFields) > 0 {
+		input.SetRemoveFields(removeFields)
+	}
+
+	_, err := conn.ModifyTrafficMirrorFilterRule(input)
+	if err != nil {
+		return fmt.Errorf("Error modifying rule %v", ruleId)
+	}
+
+	return resourceAwsTrafficMirrorFilterRuleRead(d, meta)
+}
+
+func resourceAwsTrafficMirrorFilterRuleDelete(d *schema.ResourceData, meta interface{}) error {
+	conn := meta.(*AWSClient).ec2conn
+
+	ruleId := d.Id()
+	input := &ec2.DeleteTrafficMirrorFilterRuleInput{
+		TrafficMirrorFilterRuleId: &ruleId,
+	}
+
+	_, err := conn.DeleteTrafficMirrorFilterRule(input)
+	if err != nil {
+		return fmt.Errorf("Error deleting traffic mirror filter rule %v", ruleId)
+	}
+
+	return nil
+}
+
+func buildTrafficMirrorPortRangeRequest(p []interface{}) (out *ec2.TrafficMirrorPortRangeRequest) {
+	portSchema := p[0].(map[string]interface{})
+
+	portRange := ec2.TrafficMirrorPortRangeRequest{}
+	if v, ok := portSchema["from_port"]; ok {
+		portRange.FromPort = aws.Int64(int64(v.(int)))
+		out = &portRange
+	}
+
+	if v, ok := portSchema["to_port"]; ok {
+		portRange.ToPort = aws.Int64(int64(v.(int)))
+		out = &portRange
+	}
+
+	return out
+}
+
+func populateTrafficMirrorFilterRuleResource(d *schema.ResourceData, rule *ec2.TrafficMirrorFilterRule) error {
+	d.SetId(*rule.TrafficMirrorFilterRuleId)
+	d.Set("traffic_mirror_filter_id", rule.TrafficMirrorFilterId)
+	d.Set("destination_cidr_block", rule.DestinationCidrBlock)
+	d.Set("source_cidr_block", rule.SourceCidrBlock)
+	d.Set("rule_action", rule.RuleAction)
+	d.Set("rule_number", rule.RuleNumber)
+	d.Set("traffic_direction", rule.TrafficDirection)
+	d.Set("description", rule.Description)
+	d.Set("protocol", rule.Protocol)
+	d.Set("destination_port_range", buildTrafficMirrorFilterRulePortRangeSchema(rule.DestinationPortRange))
+	d.Set("source_port_range", buildTrafficMirrorFilterRulePortRangeSchema(rule.SourcePortRange))
```

Comment:
When using `d.Set()` with aggregate types (TypeList, TypeSet, TypeMap), [we should perform error checking to prevent issues where the code is not properly able to set the Terraform state](https://www.terraform.io/docs/extend/best-practices/detecting-drift.html#error-checking-aggregate-types). e.g.

```go
if err := d.Set("destination_port_range", buildTrafficMirrorFilterRulePortRangeSchema(rule.DestinationPortRange)); err != nil {
	return fmt.Errorf("error setting destination_port_range: %s", err)
}
if err := d.Set("source_port_range", buildTrafficMirrorFilterRulePortRangeSchema(rule. SourcePortRange)); err != nil {
	return fmt.Errorf("error setting source_port_range: %s", err)
}
```